sql/query/expr.go

```diff
@@ -0,0 +1,651 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package query
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/sql/sqlwire"
+)
+
+type opType int
+
+const (
+	intOp opType = iota
+	uintOp
+	floatOp
+	stringOp
+)
+
+var null = sqlwire.Datum{}
+
+// Env defines the interface for retrieving column values.
+type Env interface {
+	Get(name string) (sqlwire.Datum, bool)
+}
+
+// mapEnv is an Env implementation using a map.
+type mapEnv map[string]sqlwire.Datum
+
+func (e mapEnv) Get(name string) (sqlwire.Datum, bool) {
+	d, ok := e[name]
+	return d, ok
+}
+
+// EvalExpr evaluates an SQL expression in the context of an
+// environment. Expression evaluation is a mostly straightforward walk over the
+// parse tree. The only significant complexity is the handling of types and
+// implicit conversions. See prepareComparisonArgs and prepareBinaryArgs for
+// details.
+func EvalExpr(expr parser.Expr, env Env) (sqlwire.Datum, error) {
+	switch t := expr.(type) {
+	case *parser.AndExpr:
+		return evalAndExpr(t, env)
+
+	case *parser.OrExpr:
+		return evalOrExpr(t, env)
+
+	case *parser.NotExpr:
+		return evalNotExpr(t, env)
+
+	case *parser.ParenBoolExpr:
+		return EvalExpr(t.Expr, env)
+
+	case *parser.ComparisonExpr:
+		return evalComparisonExpr(t, env)
+
+	case *parser.RangeCond:
+		return evalRangeCond(t, env)
+
+	case *parser.NullCheck:
+		return evalNullCheck(t, env)
+
+	case *parser.ExistsExpr:
+		// The subquery within the exists should have been executed before
+		// expression evaluation and the exists nodes replaced with the result.
+
+	case parser.BytesVal:
+		v := string(t)
+		return sqlwire.Datum{StringVal: &v}, nil
+
+	case parser.StrVal:
+		v := string(t)
+		return sqlwire.Datum{StringVal: &v}, nil
+
+	case parser.IntVal:
+		v, err := strconv.ParseUint(string(t), 0, 64)
+		if err != nil {
+			return null, err
+		}
+		return sqlwire.Datum{UintVal: &v}, nil
+
+	case parser.NumVal:
+		v, err := strconv.ParseFloat(string(t), 64)
+		if err != nil {
+			return null, err
+		}
+		return sqlwire.Datum{FloatVal: &v}, nil
+
+	case parser.BoolVal:
+		v := bool(t)
+		return sqlwire.Datum{BoolVal: &v}, nil
+
+	case parser.ValArg:
+		// Placeholders should have been replaced before expression evaluation.
+
+	case *parser.NullVal:
```

Comment:
Why is `NullVal` a pointer type instead of a value type? For that matter, it seems like many of the pointer types here can and should be values.